File: app/src/test/java/be/robinj/distrohopper/PreferencesNavigationTest.java
package be.robinj.distrohopper;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

import android.content.Intent;

import androidx.test.core.app.ActivityScenario;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.Shadows;
import org.robolectric.annotation.LooperMode;
import org.robolectric.shadows.ShadowActivity;
import org.robolectric.shadows.ShadowLooper;

import be.robinj.distrohopper.preferences.PreferencesActivity;

@RunWith(RobolectricTestRunner.class)
@LooperMode(LooperMode.Mode.LEGACY)
public class PreferencesNavigationTest {

    private ActivityScenario<HomeActivity> scenario;

    @Before
    public void setUp() {
        HomeActivityTest.seedPackageManager();
        scenario = ActivityScenario.launch(HomeActivity.class);
        Robolectric.flushBackgroundThreadScheduler();
        ShadowLooper.runUiThreadTasksIncludingDelayedTasks();
    }

    @After
    public void tearDown() {
        scenario.close();
    }

    @Test
    public void clickingCogButtonLaunchesPreferencesActivity() {
        onView(withId(R.id.ibPanelCog)).perform(click());
        scenario.onActivity(activity -> {
            ShadowActivity.IntentForResult intentForResult =
                    Shadows.shadowOf(activity).getNextStartedActivityForResult();
            assertNotNull("Cog click must start PreferencesActivity", intentForResult);
            assertEquals(PreferencesActivity.class.getName(),
                    intentForResult.intent.getComponent().getClassName());
        });
    }

    @Test
    public void homeContainerIsDisplayed() {
        onView(withId(R.id.llLauncherAndDashContainer)).check(matches(isDisplayed()));
    }

    @Test
    public void panelIsVisibleOnHomeScreen() {
        onView(withId(R.id.llPanel)).check(matches(isDisplayed()));
    }

    @Test
    public void dashIsClosedInitially() {
        onView(withId(R.id.llDash))
                .check(matches(androidx.test.espresso.matcher.ViewMatchers.withEffectiveVisibility(
                        androidx.test.espresso.matcher.ViewMatchers.Visibility.GONE)));
    }
}

File: app/src/test/java/be/robinj/distrohopper/PreferencesNavigationTest.kt
package be.robinj.distrohopper

import androidx.test.core.app.ActivityScenario
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.Visibility.GONE
import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
import androidx.test.espresso.matcher.ViewMatchers.withEffectiveVisibility
import androidx.test.espresso.matcher.ViewMatchers.withId
import be.robinj.distrohopper.preferences.PreferencesActivity
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows
import org.robolectric.annotation.LooperMode

@RunWith(RobolectricTestRunner::class)
@LooperMode(LooperMode.Mode.LEGACY)
class PreferencesNavigationTest {
    private lateinit var scenario: ActivityScenario<HomeActivity>

    @Before fun setUp() { scenario = ActivityTestSupport.launchHome() }
    @After fun tearDown() { scenario.close() }

    @Test fun clickingCogButtonLaunchesPreferencesActivity() {
        onView(withId(R.id.ibPanelCog)).perform(click())
        scenario.onActivity {
            val started = Shadows.shadowOf(it).nextStartedActivityForResult
            assertNotNull(started)
            assertEquals(PreferencesActivity::class.java.name, started.intent.component?.className)
        }
    }

    @Test fun homeContainerIsDisplayed() {
        onView(withId(R.id.llLauncherAndDashContainer)).check(matches(isDisplayed()))
    }

    @Test fun panelIsVisibleOnHomeScreen() {
        onView(withId(R.id.llPanel)).check(matches(isDisplayed()))
    }

    @Test fun dashIsClosedInitially() {
        onView(withId(R.id.llDash)).check(matches(withEffectiveVisibility(GONE)))
    }
}
